### PR TITLE
feat(hogql): Update modifiers and schemas to reflect person-on-events modes from #21455

### PR DESCRIPTION
**Pull request by tkaemming**
## Problem

#21455 clarified the various combinations of persons-on-events overrides and property modes that we want to support, so this change implements support for all the variations defined there in HogQL.

## Changes

- Updates `posthog.utils.PersonOnEventsMode` to fully remove references to "v1" and "v2" modes.
- Updates the HogQL schema `PersonsOnEventsMode` to align with the `posthog.utils.PersonOnEventsMode` values, dropping modes that do not have an equivalent value from the `PersonOnEventsMode` (e.g. the `v1_mixed` mode of `person_id` via `person_distinct_id` to `person` join and person properties is not something we want to offer, so it was removed.)
- Fixes some issues with lazy join resolution related to to the new schema (thanks @Gilbert09!)

## Does this work well for both Cloud and self-hosted?

Yes, these changes here should be backwards compatible by default with existing installations, with caveats about non-default behavior:

1. `PERSON_ID_OVERRIDE_PROPERTIES_JOINED` remains unimplemented for legacy queries, so teams should not be opted into that path if there is the potential for some of their queries to still be handled by the legacy path.
2. Self-hosted users should not opt into mode that uses overrides without having backfilled the appropriate override table (which is not currently documented.)

## How did you test this code?

Updated and added tests for schema variants, also did some ad hoc testing via the HogQL `/debug` view.

### CODE REVIEW

File: posthog/hogql/transforms/lazy_tables.py [hunk 1]
```diff
@@ -62,6 +62,9 @@ def visit_lazy_join_type(self, node: ast.LazyJoinType):
     def visit_lazy_table_type(self, node: ast.TableType):
         self.found_lazy = True
 
+    def visit_field_type(self, node: ast.FieldType):
+        self.visit(node.table_type)
```
Comment (tkaemming) [inline posthog/hogql/transforms/lazy_tables.py:66]: Well, while this change will get this test to run: https://github.com/PostHog/posthog/blob/f58b3458b05dde55ac25ebf948cdf419aa2a11cb/posthog/hogql/transforms/test/test_lazy_tables.py#L138-L147

… it seems like this creates a condition where the visitor can find itself on a cyclical path where it never stops visiting. The funnel tests demonstrate this behavior if run with

```
py.test posthog/hogql_queries/insights/funnels/test/test_funnel.py -k test_funnel_cohort_breakdown 
```

Comment (tkaemming) [reply posthog/hogql/transforms/lazy_tables.py]: Resolved by ab22a1361a90cc988976c37d941ec388c77443ff!

File: posthog/hogql/database/database.py [hunk 2]
```diff
@@ -143,6 +142,26 @@ def add_warehouse_tables(self, **field_definitions: Any):
...
+        join_table=PersonOverridesTable(),
+        join_function=join_with_person_overrides_table,
+    )
+    database.events.fields["person_id"] = ExpressionField(
+        name="person_id",
+        expr=parse_expr(
+            "ifNull(nullIf(override.override_person_id, '00000000-0000-0000-0000-000000000000'), event_person_id)",
+            start=None,
+        ),
+    )
```
Comment (tkaemming) [inline posthog/hogql/database/database.py:162]: Note that this is still using the old "v2" overrides table (which is not fully backfilled but is being kept up-to-date with realtime data) as that is the only one right now that is backwards compatible with legacy queries.

I'll follow up with a separate change that allows opting into the new "v3" table which _is_ fully backfilled (and is also kept up-to-date with realtime data), but nobody should be opted into that path that isn't either _completely_ on HogQL (for _all_ queries), _or_ okay with some queries/insights using different sets of overrides that have different data as HogQL will use v3, legacy will be stuck on v2 (how big of a discrepancy this is depends on the specific team and how often they merge/split persons.)

File: posthog/hogql/database/database.py [hunk 3]
```diff
@@ -163,52 +182,22 @@ def create_hogql_database(
...
-                start=None,
-            ),
-        )
+    elif modifiers.personsOnEventsMode == PersonsOnEventsMode.person_id_override_properties_on_events:
+        _use_person_id_from_person_overrides(database)
+        _use_person_properties_from_events(database)
         database.events.fields["poe"].fields["id"] = database.events.fields["person_id"]
-        database.events.fields["person"] = FieldTraverser(chain=["poe"])
-
-    elif modifiers.personsOnEventsMode == PersonsOnEventsMode.v3_enabled:
```
Comment (tkaemming) [inline posthog/hogql/database/database.py:195]: This branch was temporarily removed: nobody was flagged into it and #21455 made it unreachable outside of tests/debug. I'll bring it back as a follow up (see comment on `_use_person_id_from_person_overrides`.)

File: posthog/hogql/database/database.py (hunk 2 above)
Comment (tkaemming) [reply posthog/hogql/database/database.py:162]: WIP of new/v3/`person_distinct_id_overrides` support for HogQL _only_ is here: https://github.com/PostHog/posthog/compare/hogql-persons-on-events...hogql-persons-on-events-v3-overrides

**Review by timgl — APPROVED** [review]

**Pull request merged**